### DISCUSSION TITLE
vm.expectRevert(); not working as expected!!

### DISCUSSION THREAD

**Discussion by Marmik343** [Cyfrin/foundry-full-course-cu#101] [Q&A] [answered]
**This is the function  in the test file**

    function testFailsWithoutEnoughETH() public{
        vm.expectRevert(); //expect a revert to happen
        fundMe.fund();  
    }

**This is the function in the FundMe.sol contract**

```
function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
```
  
**This is not working as expected!!**  
  
```
   forge test --mt testFailsWithoutEnoughETH -vvv
[⠒] Compiling...
    ├─ [14884] FundMe::fund()
    │   ├─ [8993] MockV3Aggregator::latestRoundData() [staticcall]
    │   │   └─ ← 1, 200000000000, 1, 1, 1
    │   └─ ← "You need to spend more ETH!"
    └─ ← ()

Test result: FAILED. 0 passed; 1 failed; finished in 1.38ms

Failing tests:
Encountered 1 failing test in test/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: Assertion failed.] testFailsWithoutEnoughETH() (gas: 22945)
```

**Comment by todorchomakov** [comment]
Hi :) Can you please provide the content of your PriceConverter.sol contract. It looks like the problem you have is related to the conversion rate.

**Reply by Marmik343** [reply]
Hi, here is the code

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

library PriceConverter {
   
    function getPrice(AggregatorV3Interface priceFeed) internal view returns (uint256) {

        (, int256 answer, , , ) = priceFeed.latestRoundData();
   
        return uint256(answer * 10000000000);
    }

    function getConversionRate(
        uint256 ethAmount , AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {

        uint256 ethPrice = getPrice(priceFeed);

        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
     
        return ethAmountInUsd;
    }
}

**Reply by todorchomakov** [reply]
Everything looks correct. Maybe you have to share your code. But before doing that - make sure to double check that DeployFundMe.s.sol, HelperConfig.s.sol are all alligned to the ones in the course repo. I spoke about similar problem with another guy watching the course - he was struggling on the same part and he was using wrong constant that was breaking everything for him.

**Comment by PatrickAlphaC** [comment]
Hello! Thanks for making this. 

Could you please remember to follow this section to format your questions? https://youtu.be/umepbfKp5rI?t=22464

**Reply by Marmik343** [reply]
Indeed Patrick!
I would also like to thank you for making this course, I know for sure It will help me immensely!

**Comment by alymurtazamemon** [comment]
@Marmik343 Please run `foundryup` to update it, and can you please try this one?
```solidity
vm.expectRevert(bytes("You need to spend more ETH!"));
```

Let me know, if it does not work then please share your repository link.

**Reply by Marmik343** [reply]
https://github.com/Marmik343/Foundry__FundMe

This is my repo. I tried to run foundryup but it failed

Thanks!

**Comment by leaniegoj** [comment]
Hello! Thanks for making this. Thanks!

**Comment by alinciprian** [comment]
I have the same issue. Did anyone find a solution?
 it looks like if I use the following line, the test passes no matter what string I use, so I figure something is not working properly.

 ```Solidity
vm.expectRevert(bytes("You need to spend more ETH!"));
```


in the same time the following function test keeps failing and I can't figure out why

```Solidity
function testFailsWithoutEnoughEth() public {
        vm.expectRevert();
        fundMe.fund();
    }

```


```
[FAIL. Reason: Assertion failed.] testFailsWithoutEnoughEth() (gas: 13208)
Traces:
  [13208] FundMeTest::testFailsWithoutEnoughEth() 
    ├─ [0] VM::expectRevert() 
    │   └─ ← ()
    ├─ [5170] FundMe::fund() 
    │   ├─ [0] 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419::latestRoundData() [staticcall]
    │   │   └─ ← ()
    │   └─ ← "EvmError: Revert"
    └─ ← ()
```
The other tests are working properly.

**Reply by Equious** [reply]
This, turns out, is actually a bug in Foundry. the `expectRevert()` function is correctly identifying that a passed string != the expected error, but it still passes the test.

We've opened an issue on the [Foundry GitHub](https://github.com/foundry-rs/forge-std/issues/480)

**Comment by demhack012** [comment, marked as answer]
The problem is in the naming of the function, If your name starts with ```test``` foundry will consider it a test case but if the name starts with ```testFail``` foundry will expect the test to fail so you don't have to use ```vm.expectRevert();```

You either do your test like this:
```
    function testFundFailsWithoutEnoughEth() public {
        vm.expectRevert();
        fundMe.fund();
    }
```

or you do it like this:
```
function testFailsWithoutEnoughEth() public {
        fundMe.fund();
    }
```

both of these tests are for the same test case and will pass without any problems

Check this for more info [Writing Tests in Foundry](https://book.getfoundry.sh/forge/writing-tests#writing-tests)

@Marmik343 @todorchomakov @PatrickAlphaC @alymurtazamemon @Equious

**Reply by Equious** [reply]
Ya, ironically that's what Foundry said when we looked into it last year lol.  Updating the answer here.
